File: leetcode/230 Kth Smallest Element in a BST/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func kthSmallest(root *TreeNode, k int) int { // 方法二：递归 + 计算BST左子树节点个数！！！
    leftCount, result := count(root.Left), 0
    
    if leftCount == k-1 { // 左右出口！！！
        result = root.Val
    } else if leftCount > k-1 {
        result = kthSmallest(root.Left, k) // recursion！！！
    } else {
        result = kthSmallest(root.Right, k-leftCount-1) // recursion！！！
    }

    return result
}

func count(root *TreeNode) int { // 写辅助方法计算节点数量！！！
    if root == nil {
        return 0
    }
    return count(root.Left) + count(root.Right) + 1
}

File: leetcode/230 Kth Smallest Element in a BST/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution { // 方法一：结合 中序遍历 + BST 性质，则遍历到第k个元素即可！！！
    int count = 0;
    int result;

    public int kthSmallest(TreeNode root, int k) {
        inorder(root, k);
        return result;
    }

    public void inorder(TreeNode root, int k) { // 递归方法写中序遍历最简单！！！
        if (root == null) return; // 中序遍历：左根右！！！
        inorder(root.left, k);
        count++;
        if (count == k) {
            result = root.val;
            return;
        }
        inorder(root.right, k);
    }
}